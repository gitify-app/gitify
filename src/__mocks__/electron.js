window.Notification = function (title) {
  this.title = title;

  return {
    onclick: jest.fn(),
  };
};

window.Audio = class Audio {
  constructor(path) {
    this.path = path;
  }

  play() {}
};

window.localStorage = {
  store: {},
  getItem: function (key) {
    return this.store[key];
  },
  setItem: function (key, item) {
    this.store[key] = item;
  },
  removeItem: jest.fn(),
};

window.alert = jest.fn();

module.exports = {
  ipcRenderer: {
    send: jest.fn(),
    on: jest.fn(),
    sendSync: jest.fn(),
    invoke: jest.fn((channel, ...args) => {
      switch (channel) {
        case 'get-platform':
          return Promise.resolve('darwin');
        case 'gitify:version':
          return Promise.resolve('0.0.1');
        default:
          return Promise.reject(new Error(`Unknown channel: ${channel}`));
      }
    }),
  },
  shell: {
    openExternal: jest.fn(),
  },
};
