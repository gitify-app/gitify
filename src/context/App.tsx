import {
  type ReactNode,
  createContext,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { useInterval } from '../hooks/useInterval';
import { useNotifications } from '../hooks/useNotifications';
import {
  type AccountNotifications,
  type AuthOptions,
  type AuthState,
  type AuthTokenOptions,
  type GitifyError,
  type SettingsState,
  type Status,
  Theme,
} from '../types';
import { headNotifications } from '../utils/api/client';
import { addAccount, authGitHub, getToken, getUserData } from '../utils/auth';
import { setAutoLaunch, updateTrayTitle } from '../utils/comms';
import Constants from '../utils/constants';
import { getNotificationCount } from '../utils/notifications';
import { clearState, loadState, saveState } from '../utils/storage';
import { setTheme } from '../utils/theme';

const defaultAccounts: AuthState = {
  token: null,
  enterpriseAccounts: [],
  user: null,
};

export const defaultSettings: SettingsState = {
  participating: false,
  playSound: true,
  showNotifications: true,
  showBots: true,
  showNotificationsCountInTray: false,
  openAtStartup: false,
  theme: Theme.SYSTEM,
  detailedNotifications: false,
  markAsDoneOnOpen: false,
  showAccountHostname: false,
};

interface AppContextState {
  accounts: AuthState;
  isLoggedIn: boolean;
  login: () => void;
  loginEnterprise: (data: AuthOptions) => void;
  validateToken: (data: AuthTokenOptions) => void;
  logout: () => void;

  notifications: AccountNotifications[];
  status: Status;
  errorDetails: GitifyError;
  removeNotificationFromState: (id: string, hostname: string) => void;
  fetchNotifications: () => Promise<void>;
  markNotificationRead: (id: string, hostname: string) => Promise<void>;
  markNotificationDone: (id: string, hostname: string) => Promise<void>;
  unsubscribeNotification: (id: string, hostname: string) => Promise<void>;
  markRepoNotifications: (id: string, hostname: string) => Promise<void>;
  markRepoNotificationsDone: (id: string, hostname: string) => Promise<void>;

  settings: SettingsState;
  updateSetting: (
    name: keyof SettingsState,
    value: boolean | Theme | string | null,
  ) => void;
}

export const AppContext = createContext<Partial<AppContextState>>({});

export const AppProvider = ({ children }: { children: ReactNode }) => {
  const [accounts, setAccounts] = useState<AuthState>(defaultAccounts);
  const [settings, setSettings] = useState<SettingsState>(defaultSettings);
  const {
    fetchNotifications,
    notifications,
    errorDetails,
    status,
    removeNotificationFromState,
    markNotificationRead,
    markNotificationDone,
    unsubscribeNotification,
    markRepoNotifications,
    markRepoNotificationsDone,
  } = useNotifications();

  useEffect(() => {
    restoreSettings();
  }, []);

  useEffect(() => {
    setTheme(settings.theme as Theme);
  }, [settings.theme]);

  // biome-ignore lint/correctness/useExhaustiveDependencies: We only want fetchNotifications to be called for certain account or setting changes.
  useEffect(() => {
    fetchNotifications(accounts, settings);
  }, [
    settings.participating,
    settings.showBots,
    settings.detailedNotifications,
    accounts.token,
    accounts.enterpriseAccounts.length,
  ]);

  useInterval(() => {
    fetchNotifications(accounts, settings);
  }, Constants.FETCH_INTERVAL);

  // biome-ignore lint/correctness/useExhaustiveDependencies: We need to update tray title when settings or notifications changes.
  useEffect(() => {
    const count = getNotificationCount(notifications);

    if (settings.showNotificationsCountInTray && count > 0) {
      updateTrayTitle(count.toString());
    } else {
      updateTrayTitle();
    }
  }, [settings.showNotificationsCountInTray, notifications]);

  const updateSetting = useCallback(
    (name: keyof SettingsState, value: boolean | Theme) => {
      if (name === 'openAtStartup') {
        setAutoLaunch(value as boolean);
      }

      const newSettings = { ...settings, [name]: value };
      setSettings(newSettings);
      saveState(accounts, newSettings);
    },
    [accounts, settings],
  );

  const isLoggedIn = useMemo(() => {
    return !!accounts.token || accounts.enterpriseAccounts.length > 0;
  }, [accounts]);

  const login = useCallback(async () => {
    const { authCode } = await authGitHub();
    const { token } = await getToken(authCode);
    const hostname = Constants.DEFAULT_AUTH_OPTIONS.hostname;
    const user = await getUserData(token, hostname);
    const updatedAccounts = addAccount(accounts, token, hostname, user);
    setAccounts(updatedAccounts);
    saveState(updatedAccounts, settings);
  }, [accounts, settings]);

  const loginEnterprise = useCallback(
    async (data: AuthOptions) => {
      const { authOptions, authCode } = await authGitHub(data);
      const { token, hostname } = await getToken(authCode, authOptions);
      const updatedAccounts = addAccount(accounts, token, hostname);
      setAccounts(updatedAccounts);
      saveState(updatedAccounts, settings);
    },
    [accounts, settings],
  );

  const validateToken = useCallback(
    async ({ token, hostname }: AuthTokenOptions) => {
      await headNotifications(hostname, token);

      const user = await getUserData(token, hostname);
      const updatedAccounts = addAccount(accounts, token, hostname, user);
      setAccounts(updatedAccounts);
      saveState(updatedAccounts, settings);
    },
    [accounts, settings],
  );

  const logout = useCallback(() => {
    setAccounts(defaultAccounts);
    clearState();
  }, []);

  const restoreSettings = useCallback(() => {
    const existing = loadState();

    if (existing.accounts) {
      setAccounts({ ...defaultAccounts, ...existing.accounts });
    }

    if (existing.settings) {
      setSettings({ ...defaultSettings, ...existing.settings });
      return existing.settings;
    }
  }, []);

  const fetchNotificationsWithAccounts = useCallback(
    async () => await fetchNotifications(accounts, settings),
    [accounts, settings, notifications],
  );

  const markNotificationReadWithAccounts = useCallback(
    async (id: string, hostname: string) =>
      await markNotificationRead(accounts, id, hostname),
    [accounts, notifications],
  );

  const markNotificationDoneWithAccounts = useCallback(
    async (id: string, hostname: string) =>
      await markNotificationDone(accounts, id, hostname),
    [accounts, notifications],
  );

  const unsubscribeNotificationWithAccounts = useCallback(
    async (id: string, hostname: string) =>
      await unsubscribeNotification(accounts, id, hostname),
    [accounts, notifications],
  );

  const markRepoNotificationsWithAccounts = useCallback(
    async (repoSlug: string, hostname: string) =>
      await markRepoNotifications(accounts, repoSlug, hostname),
    [accounts, notifications],
  );

  const markRepoNotificationsDoneWithAccounts = useCallback(
    async (repoSlug: string, hostname: string) =>
      await markRepoNotificationsDone(accounts, repoSlug, hostname),
    [accounts, notifications],
  );

  return (
    <AppContext.Provider
      value={{
        accounts,
        isLoggedIn,
        login,
        loginEnterprise,
        validateToken,
        logout,

        notifications,
        status,
        errorDetails,
        removeNotificationFromState,
        fetchNotifications: fetchNotificationsWithAccounts,
        markNotificationRead: markNotificationReadWithAccounts,
        markNotificationDone: markNotificationDoneWithAccounts,
        unsubscribeNotification: unsubscribeNotificationWithAccounts,
        markRepoNotifications: markRepoNotificationsWithAccounts,
        markRepoNotificationsDone: markRepoNotificationsDoneWithAccounts,

        settings,
        updateSetting,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
