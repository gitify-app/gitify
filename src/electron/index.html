<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Gitify</title>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta http-equiv="Content-Security-Policy" content="script-src 'self';" />
  </head>

  <body>
    <div
      id="gitify"
      class="bg-white text-black dark:bg-gray-dark dark:text-white"
    ></div>
  </body>

  <script src="./build/js/app.js"></script>

  <style>
    html,
    body,
    #gitify {
      height: 100%;
      -webkit-user-select: none;
    }

    html::-webkit-scrollbar,
    div.flex-grow::-webkit-scrollbar {
        width: 10px;
    }

    html::-webkit-scrollbar-thumb,
    div.flex-grow::-webkit-scrollbar-thumb {
        background-color: #c1c1c1;
        border-radius: 10px;
    }

    html::-webkit-scrollbar-thumb:hover,
    div.flex-grow::-webkit-scrollbar-thumb:hover {
        background-color: #a8a8a8;
        border-radius: 10px;
    }

    html.dark::-webkit-scrollbar-track,
    html.dark div.flex-grow::-webkit-scrollbar-track {
        background-color: #090E15;
    }

    html.dark::-webkit-scrollbar-thumb,
    html.dark div.flex-grow::-webkit-scrollbar-thumb {
        background-color: #24292e;
    }

    html.dark::-webkit-scrollbar-thumb:hover,
    html.dark div.flex-grow::-webkit-scrollbar-thumb:hover {
        background-color: #3a3f44;
    }

    body {
      font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto",
        "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans",
        "Helvetica Neue", sans-serif;
      margin: 0;
      cursor: default;
    }

    #nprogress {
      pointer-events: none;
    }

    #nprogress .bar {
      position: fixed;
      top: 0;
      left: 50px;
      z-index: 1031;
      background: #203354;

      width: 100%;
      height: 2px;
    }
  </style>
</html>
