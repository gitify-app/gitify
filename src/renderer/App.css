html,
body,
#root {
  height: 100%;
  -webkit-user-select: none;
}

html::-webkit-scrollbar,
div.flex-grow::-webkit-scrollbar {
  width: 10px;
}

html::-webkit-scrollbar-thumb,
div.flex-grow::-webkit-scrollbar-thumb {
  background-color: #c1c1c1;
  border-radius: 10px;
}

html::-webkit-scrollbar-thumb:hover,
div.flex-grow::-webkit-scrollbar-thumb:hover {
  background-color: #a8a8a8;
  border-radius: 10px;
}

html.dark::-webkit-scrollbar-track,
html.dark div.flex-grow::-webkit-scrollbar-track {
  background-color: #090e15;
}

html.dark::-webkit-scrollbar-thumb,
html.dark div.flex-grow::-webkit-scrollbar-thumb {
  background-color: #24292e;
}

html.dark::-webkit-scrollbar-thumb:hover,
html.dark div.flex-grow::-webkit-scrollbar-thumb:hover {
  background-color: #3a3f44;
}

body {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
    "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
    sans-serif;
  margin: 0;
  cursor: default;
}

#nprogress {
  pointer-events: none;
}

#nprogress .bar {
  position: fixed;
  top: 0;
  left: 50px;
  z-index: 1031;
  background: #203354;

  width: 100%;
  height: 2px;
}

/**
 * Set emoji size according to surrounding text.
 * Ref: https://github.com/jdecked/twemoji?tab=readme-ov-file#inline-styles
 */
img.emoji {
  height: 1em;
  width: 1em;
  margin: 0 0.05em 0 0.1em;
  vertical-align: -0.1em;
}
