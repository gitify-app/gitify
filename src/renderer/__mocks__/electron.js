const { namespacedEvent } = require('../../shared/events');

// @ts-ignore
window.Notification = function (title) {
  this.title = title;

  return {
    onclick: vi.fn(),
  };
};

// @ts-ignore
window.Audio = class Audio {
  constructor(path) {
    this.path = path;
  }

  play() {}
};

// @ts-ignore
window.localStorage = {
  store: {},
  getItem: function (key) {
    return this.store[key];
  },
  setItem: function (key, item) {
    this.store[key] = item;
  },
  removeItem: vi.fn(),
};

window.alert = vi.fn();

module.exports = {
  ipcRenderer: {
    send: vi.fn(),
    on: vi.fn(),
    sendSync: vi.fn(),
    invoke: vi.fn((channel, ..._args) => {
      switch (channel) {
        case 'get-platform':
          return Promise.resolve('darwin');
        case namespacedEvent('version'):
          return Promise.resolve('0.0.1');
        case namespacedEvent('safe-storage-encrypt'):
          return Promise.resolve('encrypted');
        case namespacedEvent('safe-storage-decrypt'):
          return Promise.resolve('decrypted');
        default:
          return Promise.reject(new Error(`Unknown channel: ${channel}`));
      }
    }),
  },
  shell: {
    openExternal: vi.fn(),
  },
  webFrame: {
    setZoomLevel: vi.fn(),
    getZoomLevel: vi.fn(),
  },
};
