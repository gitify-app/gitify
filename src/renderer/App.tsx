import { useContext } from 'react';
import {
  Navigate,
  Route,
  HashRouter as Router,
  Routes,
  useLocation,
} from 'react-router-dom';

import { BaseStyles, Box, ThemeProvider } from '@primer/react';

import { Loading } from './components/Loading';
import { Sidebar } from './components/Sidebar';
import { AppContext, AppProvider } from './context/App';
import { AccountsRoute } from './routes/Accounts';
import { FiltersRoute } from './routes/Filters';
import { LoginRoute } from './routes/Login';
import { LoginWithOAuthAppRoute } from './routes/LoginWithOAuthApp';
import { LoginWithPersonalAccessTokenRoute } from './routes/LoginWithPersonalAccessToken';
import { NotificationsRoute } from './routes/Notifications';
import { SettingsRoute } from './routes/Settings';

import './App.css';

function RequireAuth({ children }) {
  const { isLoggedIn } = useContext(AppContext);
  const location = useLocation();

  return isLoggedIn ? (
    children
  ) : (
    <Navigate to="/login" replace state={{ from: location }} />
  );
}

export const App = () => {
  return (
    <ThemeProvider>
      <BaseStyles>
        <AppProvider>
          <Router>
            <Box className="flex flex-col min-h-screen overflow-x-hidden overflow-y-auto pl-sidebar bg-gitify-background">
              <Loading />
              <Sidebar />
              <Routes>
                <Route
                  path="/"
                  element={
                    <RequireAuth>
                      <NotificationsRoute />
                    </RequireAuth>
                  }
                />
                <Route
                  path="/filters"
                  element={
                    <RequireAuth>
                      <FiltersRoute />
                    </RequireAuth>
                  }
                />
                <Route
                  path="/settings"
                  element={
                    <RequireAuth>
                      <SettingsRoute />
                    </RequireAuth>
                  }
                />
                <Route
                  path="/accounts"
                  element={
                    <RequireAuth>
                      <AccountsRoute />
                    </RequireAuth>
                  }
                />
                <Route path="/login" element={<LoginRoute />} />
                <Route
                  path="/login-personal-access-token"
                  element={<LoginWithPersonalAccessTokenRoute />}
                />
                <Route
                  path="/login-oauth-app"
                  element={<LoginWithOAuthAppRoute />}
                />
              </Routes>
            </Box>
          </Router>
        </AppProvider>
      </BaseStyles>
    </ThemeProvider>
  );
};
