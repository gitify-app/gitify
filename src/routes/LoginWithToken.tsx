import React, { useCallback, useContext, useState } from 'react';
import { Form, FormRenderProps } from 'react-final-form';
import { ArrowLeftIcon } from '@primer/octicons-react';
import { useHistory } from 'react-router-dom';
import { shell } from 'electron';

import { AppContext } from '../context/App';
import { AuthTokenOptions } from '../types';
import { Constants } from '../utils/constants';
import { FieldInput } from '../components/fields/FieldInput';

interface IValues {
  token?: string;
  hostname?: string;
}

interface IFormErrors {
  token?: string;
  hostname?: string;
}

export const validate = (values: IValues): IFormErrors => {
  const errors: IFormErrors = {};
  if (!values.token) {
    errors.token = 'Required';
  } else if (!/^[A-Z0-9_]{40}$/i.test(values.token)) {
    errors.token = 'Invalid token.';
  }

  if (!values.hostname) {
    errors.hostname = 'Required';
  } else if (
    !/^([a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]{2,6}$/i.test(
      values.hostname
    )
  ) {
    errors.hostname = 'Invalid hostname.';
  }

  return errors;
};

export const LoginWithToken: React.FC = () => {
  const { validateToken } = useContext(AppContext);
  const history = useHistory();
  const [isValidToken, setIsValidToken] = useState<boolean>(true);

  const openLink = useCallback((url: string) => {
    shell.openExternal(url);
  }, []);

  const renderForm = (formProps: FormRenderProps) => {
    const { handleSubmit, submitting, pristine, values } = formProps;

    return (
      <form onSubmit={handleSubmit}>
        <FieldInput
          name="token"
          label="Token"
          placeholder="The 40 characters token generated on GitHub"
          helpText={
            <>
              To generate a token, go to GitHub,{' '}
              <a
                className="underline hover:text-gray-500 dark:hover:text-gray-300  cursor-pointer"
                onClick={() => openLink('https://github.com/settings/tokens')}
              >
                personal access tokens
              </a>{' '}
              and create one with the{' '}
              <span className="underline font-extrabold text-yellow-500">
                {Constants.AUTH_SCOPE.join(', ')}{' '}
              </span>
              scopes.
            </>
          }
        />

        <FieldInput
          name="hostname"
          label="Hostname"
          placeholder="github.company.com"
          helpText="Defaults to github.com. Change only if you are using GitHub for Enterprise."
        />

        {!isValidToken && (
          <div className="mt-4 text-red-500 text-sm font-medium">
            This token could not be validated with {values.hostname}.
          </div>
        )}

        <button
          className="float-right px-4 py-2 my-4 bg-gray-300 font-semibold rounded text-sm text-center hover:bg-gray-500 hover:text-white dark:text-black focus:outline-none"
          disabled={submitting || pristine}
          type="submit"
          title="Submit Button"
        >
          Submit
        </button>
      </form>
    );
  };

  const submit = useCallback(async (data: IValues) => {
    setIsValidToken(true);
    try {
      await validateToken(data as AuthTokenOptions);
      history.goBack();
    } catch (err) {
      setIsValidToken(false);
    }
  }, []);

  return (
    <div className="flex-1 bg-white dark:bg-gray-dark dark:text-white">
      <div className="flex justify-between items-center mt-4 py-2 mx-8">
        <button
          className="focus:outline-none"
          aria-label="Go Back"
          onClick={() => history.goBack()}
        >
          <ArrowLeftIcon size={20} className="hover:text-gray-400" />
        </button>

        <h3 className="text-lg font-semibold">Login with an access token</h3>
      </div>

      <div className="flex-1 px-8">
        <Form
          initialValues={{
            token: '',
            hostname: Constants.DEFAULT_AUTH_OPTIONS.hostname,
          }}
          onSubmit={submit}
          validate={validate}
        >
          {renderForm}
        </Form>
      </div>
    </div>
  );
};
