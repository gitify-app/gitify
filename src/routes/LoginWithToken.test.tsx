import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import { shell } from 'electron';

import { MemoryRouter } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';

import { AppContext } from '../context/App';
import { LoginWithToken, validate } from './LoginWithToken';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('routes/LoginWithToken.tsx', () => {
  const openExternalMock = jest.spyOn(shell, 'openExternal');

  const mockValidateToken = jest.fn();

  beforeEach(() => {
    mockValidateToken.mockReset();
    openExternalMock.mockReset();
    mockNavigate.mockReset();
  });

  it('renders correctly', () => {
    const tree = TestRenderer.create(
      <MemoryRouter>
        <LoginWithToken />
      </MemoryRouter>,
    );

    expect(tree).toMatchSnapshot();
  });

  it('let us go back', () => {
    render(
      <MemoryRouter>
        <LoginWithToken />
      </MemoryRouter>,
    );

    fireEvent.click(screen.getByLabelText('Go Back'));
    expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
  });

  it('should validate the form values', () => {
    const emptyValues = {
      hostname: null,
      token: null,
    };

    let values = {
      ...emptyValues,
    };
    expect(validate(values).hostname).toBe('Required');
    expect(validate(values).token).toBe('Required');

    values = {
      ...emptyValues,
      hostname: 'hello',
      token: '!@Â£INVALID-.1',
    };
    expect(validate(values).hostname).toBe('Invalid hostname.');
    expect(validate(values).token).toBe('Invalid token.');
  });

  it("should click on the 'Generate a PAT' link and open the browser", async () => {
    render(
      <AppContext.Provider value={{ validateToken: mockValidateToken }}>
        <MemoryRouter>
          <LoginWithToken />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.click(screen.getByText('Generate a PAT'));

    expect(openExternalMock).toHaveBeenCalledTimes(1);
  });

  it('should login using a token - success', async () => {
    mockValidateToken.mockResolvedValueOnce(null);

    render(
      <AppContext.Provider value={{ validateToken: mockValidateToken }}>
        <MemoryRouter>
          <LoginWithToken />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    fireEvent.change(screen.getByLabelText('Token'), {
      target: { value: '1234567890123456789012345678901234567890' },
    });
    fireEvent.change(screen.getByLabelText('Hostname'), {
      target: { value: 'github.com' },
    });

    fireEvent.submit(screen.getByTitle('Submit Button'));

    await waitFor(() => expect(mockValidateToken).toHaveBeenCalledTimes(1));

    expect(mockValidateToken).toHaveBeenCalledTimes(1);
    expect(mockNavigate).toHaveBeenNthCalledWith(1, -1);
  });

  it('should login using a token - failure', async () => {
    mockValidateToken.mockRejectedValueOnce(null);

    render(
      <AppContext.Provider value={{ validateToken: mockValidateToken }}>
        <MemoryRouter>
          <LoginWithToken />
        </MemoryRouter>
      </AppContext.Provider>,
    );

    act(() => {
      fireEvent.change(screen.getByLabelText('Token'), {
        target: { value: '1234567890123456789012345678901234567890' },
      });
      fireEvent.change(screen.getByLabelText('Hostname'), {
        target: { value: 'github.com' },
      });
      fireEvent.submit(screen.getByTitle('Submit Button'));
    });

    await waitFor(() => expect(mockValidateToken).toHaveBeenCalledTimes(1));

    expect(mockValidateToken).toHaveBeenCalledTimes(1);
    expect(mockNavigate).toHaveBeenCalledTimes(0);
  });

  it('should render the form with errors', () => {
    render(
      <MemoryRouter>
        <LoginWithToken />
      </MemoryRouter>,
    );

    fireEvent.change(screen.getByLabelText('Hostname'), {
      target: { value: 'test' },
    });
    fireEvent.change(screen.getByLabelText('Token'), {
      target: { value: '123' },
    });

    fireEvent.submit(screen.getByTitle('Submit Button'));

    expect(screen.getByText('Invalid hostname.')).toBeDefined();
    expect(screen.getByText('Invalid token.')).toBeDefined();
  });
});
