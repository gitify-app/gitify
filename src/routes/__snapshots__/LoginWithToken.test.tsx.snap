// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes/LoginWithToken.js renders correctly 1`] = `
<div
  className="flex-1 bg-white dark:bg-gray-dark dark:text-white"
>
  <div
    className="flex justify-between items-center mt-4 py-2 mx-8"
  >
    <button
      aria-label="Go Back"
      className="focus:outline-none"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="hover:text-gray-400"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path fill-rule=\\"evenodd\\" d=\\"M7.78 12.53a.75.75 0 01-1.06 0L2.47 8.28a.75.75 0 010-1.06l4.25-4.25a.75.75 0 011.06 1.06L4.81 7h7.44a.75.75 0 010 1.5H4.81l2.97 2.97a.75.75 0 010 1.06z\\"></path>",
          }
        }
        fill="currentColor"
        height={20}
        role="img"
        style={
          Object {
            "display": "inline-block",
            "userSelect": "none",
            "verticalAlign": "text-bottom",
          }
        }
        viewBox="0 0 16 16"
        width={20}
      />
    </button>
    <h3
      className="text-lg font-semibold"
    >
      Login with an access token
    </h3>
  </div>
  <div
    className="flex-1 px-8"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="mt-2"
      >
        <label
          className="block tracking-wide text-grey-dark text-sm font-semibold mb-2"
          htmlFor="token"
        >
          Token
        </label>
        <input
          className="appearance-none block w-full dark:text-gray-800 bg-gray-100 border border-red rounded py-1.5 px-4 mb-2 focus:bg-gray-200 focus:outline-none"
          id="token"
          name="token"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="The 40 characters token generated on GitHub"
          required={false}
          type="text"
          value=""
        />
        <div
          className="mt-3 text-gray-700 dark:text-gray-200 text-xs"
        >
          To generate a token, go to GitHub,
           
          <a
            className="underline hover:text-gray-500 dark:hover:text-gray-300  cursor-pointer"
            onClick={[Function]}
          >
            personal access tokens
          </a>
           
          and create one with the
           
          <span
            className="underline font-extrabold text-yellow-500"
          >
            read:user, notifications
             
          </span>
          scopes.
        </div>
      </div>
      <div
        className="mt-2"
      >
        <label
          className="block tracking-wide text-grey-dark text-sm font-semibold mb-2"
          htmlFor="hostname"
        >
          Hostname
        </label>
        <input
          className="appearance-none block w-full dark:text-gray-800 bg-gray-100 border border-red rounded py-1.5 px-4 mb-2 focus:bg-gray-200 focus:outline-none"
          id="hostname"
          name="hostname"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="github.company.com"
          required={false}
          type="text"
          value="github.com"
        />
        <div
          className="mt-3 text-gray-700 dark:text-gray-200 text-xs"
        >
          Defaults to github.com. Change only if you are using GitHub for Enterprise.
        </div>
      </div>
      <button
        className="float-right px-4 py-2 my-4 bg-gray-300 font-semibold rounded text-sm text-center hover:bg-gray-500 hover:text-white dark:text-black focus:outline-none"
        disabled={true}
        title="Submit Button"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
